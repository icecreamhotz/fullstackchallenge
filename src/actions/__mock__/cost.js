import { UPDATE_COST, UPDATE_MONEY, DELETE_MONEY } from "../../types";

export const setCost = cost => dispatch => {
  dispatch({
    type: UPDATE_COST,
    payload: {
      cost: {
        _id: "5ceea163de8adf2ef81d502b",
        locker: 1,
        size: {
          _id: "5cee905b5162c31b700a185a",
          size: "S"
        },
        income: 700,
        status: "1",
        timeout: "2019-06-02T18:49:29.166",
        user: {
          _id: "5cf016743caf73157c995336",
          telephone: "0850300073"
        }
      }
    }
  });
};

export const setMoney = money => dispatch => {
  dispatch({
    type: UPDATE_MONEY,
    payload: { money: money }
  });
};

export const deleteMoney = money => dispatch => {
  dispatch({
    type: DELETE_MONEY,
    payload: { money: 0 }
  });
};
