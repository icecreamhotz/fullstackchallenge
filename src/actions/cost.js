import { UPDATE_COST, UPDATE_MONEY, DELETE_MONEY } from "../types";

export const setCost = cost => dispatch => {
  dispatch({
    type: UPDATE_COST,
    cost
  });
};

export const setMoney = money => dispatch => {
  dispatch({
    type: UPDATE_MONEY,
    money
  });
};

export const deleteMoney = money => dispatch => {
  dispatch({
    type: DELETE_MONEY,
    money
  });
};
