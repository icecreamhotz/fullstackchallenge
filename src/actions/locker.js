import { UPDATE_LOCKER_SELECTED } from "../types";

export const setLockers = lockers => dispatch => {
  dispatch({
    type: UPDATE_LOCKER_SELECTED,
    lockers
  });
};
