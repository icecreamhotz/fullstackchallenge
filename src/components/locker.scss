$full-width: 100%;
$bg-available: #7aff51;
$bg-unavailable: #ed4325;
$bg-selected: rgba(0,0,0,.125);


.align-right {
    text-align: right;
}

.align-center {
    text-align: center;
}

.pay-machine {
    height: auto;
}

.tv-touchscreen {
    width: $full-width;
    border: 1px solid rgba(0,0,0,.125);
    padding: 5px;
}

.content-line-header, .content-line-common {
    width: $full-width;
    margin-bottom: 16px;
    border-radius: 8px;
}

.content-line-header.header {
    margin: 10px auto 20px auto;
    height: 12px;
}

.content-line-common.common {
    height: 10px;
}

.content-line-common.price {
    height: 8px;
}

.endheader {
    width: 80%;
}

.endcommon {
    width: 95%;
}

.endcoin-header {
    width: 40%;
}

.shimmerBG {
    animation-duration: 2.2s;
    animation-fill-mode: forwards;
    animation-iteration-count: infinite;
    animation-name: shimmer;
    animation-timing-function: linear;
    background: #d9d9d9;
    background: linear-gradient(to right, #ecebeb 4%, #d9d9d9 25%, #ecebeb 36%);
    background-size: 1200px 100%;
}

@-webkit-keyframes shimmer {
    0% {
        background-position: -100% 0;
    }
    100% {
        background-position: 100% 0;
    }
}

@keyframes shimmer {
    0% {
        background-position: -1200px 0;
    }
    100% {
        background-position: 1200px 0;
    }
}

.unit {
    width: 80%;
    border-radius: 10px;
    text-align: center;
    padding-top: 15px;
    padding-bottom: 15px;
    margin: 0 auto 10px auto;

    &:hover {
        cursor: pointer;
    }
}

.cursor-unavailable {
    &:hover {
        cursor: not-allowed;
    }
}

.available {
    background-color: $bg-available;
}

.unavailable {
    background-color: $bg-unavailable;
}

.selected {
    background-color: $bg-selected;
}

.rended {
    background-color: #ed4325c7;
}