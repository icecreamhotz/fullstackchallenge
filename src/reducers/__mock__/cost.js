import { UPDATE_COST, UPDATE_MONEY, DELETE_MONEY } from "../../types";

export default function cost(state = { cost: {}, money: 0 }, action = {}) {
  switch (action.type) {
    case UPDATE_COST:
      return Object.assign({}, state, {
        cost: action.payload.cost
      });
    case UPDATE_MONEY:
      return Object.assign({}, state, {
        money: action.payload.money
      });
    case DELETE_MONEY:
      return Object.assign({}, state, {
        money: 0
      });
    default:
      return state;
  }
}
