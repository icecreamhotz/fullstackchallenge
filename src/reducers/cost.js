import { UPDATE_COST, UPDATE_MONEY, DELETE_MONEY } from "../types";

export default function cost(state = { cost: {}, money: 0 }, action = {}) {
  switch (action.type) {
    case UPDATE_COST:
      return { ...state, cost: action.cost };
    case UPDATE_MONEY:
      return {
        ...state,
        money: action.money + state.money
      };
    case DELETE_MONEY:
      return {
        ...state,
        money: 0
      };
    default:
      return state;
  }
}
