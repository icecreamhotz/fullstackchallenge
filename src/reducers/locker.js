import { UPDATE_LOCKER_SELECTED } from "../types";

export default function locker(state = {}, action = {}) {
  switch (action.type) {
    case UPDATE_LOCKER_SELECTED:
      return action.lockers;
    default:
      return state;
  }
}
