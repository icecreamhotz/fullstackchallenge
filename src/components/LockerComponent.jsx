import React, { Component } from "react";
import { Container, Row, Col } from "react-bootstrap";
import "./locker.scss";
import { connect } from "react-redux";
import { setLockers } from "../actions/locker.js";

import lockerService from "../services/locker.js";
import { socket } from "../config/socket.js";
import UnitsComponent from "./UnitsComponent";
import MachineComponent from "./MachineComponent";

class LockerComponent extends Component {

  state = { loading: true }

  async componentDidMount() {
    await this.fetchLockers();
    socket.on("lockers", data => {
      const addSelectedKey = this.mapLocker(data);
      this.props.setLockers(addSelectedKey);
    });
  }

  fetchLockers = async () => {
    try {
      const locker = await lockerService.getAllLocker();
      const lockerData = await locker.data.lockers;
      const addSelectedKey = this.mapLocker(lockerData);
      this.props.setLockers(addSelectedKey);
      this.setState({
        loading: false
      })
    } catch (err) {
      alert("something has wrong");
      console.log(err);
    }
  };

  mapLocker = data => {
    return data.map(locker => {
      return {
        ...locker,
        selected: "0"
      };
    });
  };

  render() {
    const { loading } = this.state;
    return (
      <div style={{ paddingTop: 30 }}>
        <Container>
          <Row>
            <Col xs={3}>
              <MachineComponent loading={loading} />
            </Col>
            <Col xs={9}>
              <UnitsComponent loading={loading} />
            </Col>
          </Row>
        </Container>
      </div>
    );
  }
}

export default connect(
  null,
  { setLockers: setLockers }
)(LockerComponent);
