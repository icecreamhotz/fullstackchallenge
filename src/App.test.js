import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";

import LockerComponent from "./components/LockerComponent.jsx";

import { mockLockers } from "./services/__mocks__/locker";
import { setCost, setMoney, deleteMoney } from "./actions/cost";

import cost from "./reducers/__mock__/cost";

import {
  setStatusSuccessModal,
  setStatusUnAvailableModal,
  setChangeToSuccessModal
} from "./actions/modal";

const mockStore = configureMockStore([thunk]);

Enzyme.configure({ adapter: new Adapter() });

describe("Test Case 1", () => {
  let store;
  let costReduce;

  beforeEach(() => {
    // mock data
    store = mockStore({
      locker: [],
      cost: { cost: {}, money: 0 },
      modal: { successModal: false, unAvailableModal: false, change: 0 },
      loading: false
    });
  });

  test("renders", () => {
    // check component has exists
    const wrapper = shallow(
      <Provider store={store}>
        <LockerComponent />
      </Provider>
    );

    expect(wrapper.exists()).toBe(true);
  });

  test("Request Locker", async () => {
    // mock api
    await store.dispatch(mockLockers());
    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_LOCKER_SELECTED");
  });

  test("Set money and select locker", async () => {
    // mock insert cash 1000
    await store.dispatch(setMoney(1000));
    // mock locker
    await store.dispatch(setCost());
    const actions = store.getActions();
    const costs = cost(undefined, {
      type: "UPDATE_MONEY",
      payload: {
        money: 1000
      }
    });
    // expect insert cash 1000 success
    expect(costs.money).toEqual(1000);
    // mock locker
    costReduce = cost(undefined, {
      type: "UPDATE_COST",
      payload: {
        cost: {
          _id: "5ceea163de8adf2ef81d502b",
          locker: 1,
          size: {
            _id: "5cee905b5162c31b700a185a",
            size: "S",
            perhour: 50,
            nextminute: 25
          },
          income: 700,
          status: "1",
          timeout: "2019-06-02T18:49:29.166",
          user: {
            _id: "5cf016743caf73157c995336",
            telephone: "0850300073"
          },
          time: 60,
          change: 950
        }
      }
    });
    //expect time is 60 minutes
    expect(costReduce.cost.time).toEqual(60);
    //expect charge is 50 bath perhour
    expect(costReduce.cost.size.perhour).toEqual(50);
    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MONEY");
    //expect action redux
    expect(actions[1].type).toEqual("UPDATE_COST");
  });

  test("Confirm Order And Rent Success", async () => {
    //mock open success modal
    await store.dispatch(setStatusSuccessModal(true));
    //call function calculateChange like cash and coin
    const change = calculateChangeCashAndCoin(costReduce.cost.change);
    //expect get five hundred cash
    expect(change[1]).toEqual(1);
    //expect get one hundred cash
    expect(change[2]).toEqual(4);
    //expect get fifty cash
    expect(change[3]).toEqual(1);

    await store.dispatch(setChangeToSuccessModal(0));
    await store.dispatch(deleteMoney());

    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MODAL_SUCCESS");
    expect(actions[1].type).toEqual("UPDATE_CHANGE_SUCCESS");
    expect(actions[2].type).toEqual("DELETE_MONEY");
  });
});

describe("Test Case 2", () => {
  let store;
  let costReduce;

  beforeEach(() => {
    // mock data
    store = mockStore({
      locker: [],
      cost: { cost: {}, money: 0 },
      modal: { successModal: false, unAvailableModal: false, change: 0 },
      loading: false
    });
  });

  test("renders", () => {
    // check component has exists
    const wrapper = shallow(
      <Provider store={store}>
        <LockerComponent />
      </Provider>
    );

    expect(wrapper.exists()).toBe(true);
  });

  test("Request Locker", async () => {
    // mock api
    await store.dispatch(mockLockers());
    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_LOCKER_SELECTED");
  });

  test("Set money and select locker", async () => {
    // mock insert cash 100
    await store.dispatch(setMoney(100));
    // mock locker
    await store.dispatch(setCost());
    const actions = store.getActions();
    const costs = cost(undefined, {
      type: "UPDATE_MONEY",
      payload: {
        money: 100
      }
    });
    // expect insert cash 100 success
    expect(costs.money).toEqual(100);
    // mock locker
    costReduce = cost(undefined, {
      type: "UPDATE_COST",
      payload: {
        cost: {
          _id: "5ceea163de8adf2ef81d502b",
          locker: 1,
          size: {
            _id: "5cee905b5162c31b700a185a",
            size: "S",
            perhour: 50,
            nextminute: 25
          },
          income: 700,
          status: "1",
          timeout: "2019-06-02T18:49:29.166",
          user: {
            _id: "5cf016743caf73157c995336",
            telephone: "0850300073"
          },
          time: 61,
          change: 25
        }
      }
    });
    //expect time is 60 minutes
    expect(costReduce.cost.time).toEqual(61);
    //expect charge is 50 bath perhour
    expect(costReduce.cost.size.perhour).toEqual(50);
    //expect next minute
    expect(costReduce.cost.size.nextminute).toEqual(25);
    //expect total perhour and nextminute
    expect(
      parseInt(costReduce.cost.size.perhour + costReduce.cost.size.nextminute)
    ).toEqual(75);
    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MONEY");
    //expect action redux
    expect(actions[1].type).toEqual("UPDATE_COST");
  });

  test("Confirm Order And Rent Success", async () => {
    //mock open success modal
    await store.dispatch(setStatusSuccessModal(true));
    //call function calculateChange like cash and coin
    const change = calculateChangeCashAndCoin(costReduce.cost.change);
    //expect get twenty cash
    expect(change[4]).toEqual(1);
    //expect get one bath coin
    expect(change[6]).toEqual(1);

    await store.dispatch(setChangeToSuccessModal(0));
    await store.dispatch(deleteMoney());

    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MODAL_SUCCESS");
    expect(actions[1].type).toEqual("UPDATE_CHANGE_SUCCESS");
    expect(actions[2].type).toEqual("DELETE_MONEY");
  });
});

describe("Test Case 3", () => {
  let store;
  let costReduce;

  beforeEach(() => {
    // mock data
    store = mockStore({
      locker: [],
      cost: { cost: {}, money: 0 },
      modal: { successModal: false, unAvailableModal: false, change: 0 },
      loading: false
    });
  });

  test("renders", () => {
    // check component has exists
    const wrapper = shallow(
      <Provider store={store}>
        <LockerComponent />
      </Provider>
    );

    expect(wrapper.exists()).toBe(true);
  });

  test("Request Locker", async () => {
    // mock api
    await store.dispatch(mockLockers());
    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_LOCKER_SELECTED");
  });

  test("Set money and select locker", async () => {
    // mock insert cash 100
    await store.dispatch(setMoney(100));
    // mock locker
    await store.dispatch(setCost());
    const actions = store.getActions();
    const costs = cost(undefined, {
      type: "UPDATE_MONEY",
      payload: {
        money: 100
      }
    });
    // expect insert cash 100 success
    expect(costs.money).toEqual(100);
    // mock locker
    costReduce = cost(undefined, {
      type: "UPDATE_COST",
      payload: {
        cost: {
          _id: "5ceea163de8adf2ef81d5026",
          locker: 2,
          size: {
            _id: "5cee905b5162c31b700a185b",
            size: "M",
            perhour: 100,
            nextminute: 50
          },
          income: 700,
          status: "0",
          timeout: "2019-06-02T18:49:29.166",
          user: {
            _id: "5cf016743caf73157c995336",
            telephone: "0850300073"
          },
          time: 0,
          change: 100
        }
      }
    });
    //expect time is 120 minutes
    expect(costReduce.cost.time).toEqual(120);
    //expect charge is 100 bath perhour
    expect(costReduce.cost.size.perhour).toEqual(100);
    //expect next minute
    expect(costReduce.cost.size.nextminute).toEqual(50);
    //expect total perhour and nextminute because user insert 100 bath but machine need 150 bath
    expect(
      costReduce.cost.size.persour + costReduce.cost.size.nextminute
    ).toEqual(100);
    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MONEY");
    //expect action redux
    expect(actions[1].type).toEqual("UPDATE_COST");
  });
});

describe("Test Case 4", () => {
  let store;
  let costReduce;

  beforeEach(() => {
    // mock data
    store = mockStore({
      locker: [],
      cost: { cost: {}, money: 0 },
      modal: { successModal: false, unAvailableModal: false, change: 0 },
      loading: false
    });
  });

  test("renders", () => {
    // check component has exists
    const wrapper = shallow(
      <Provider store={store}>
        <LockerComponent />
      </Provider>
    );

    expect(wrapper.exists()).toBe(true);
  });

  test("Request Locker", async () => {
    // mock api
    await store.dispatch(mockLockers());
    const actions = store.getActions();

    costReduce = cost(undefined, {
      type: "UPDATE_COST",
      payload: {
        cost: {
          _id: "5ceea163de8adf2ef81d5026",
          locker: 2,
          size: {
            _id: "5cee905b5162c31b700a185b",
            size: "M",
            perhour: 100,
            nextminute: 50
          },
          income: 700,
          status: "1",
          timeout: "2019-06-02T18:49:29.166",
          user: {
            _id: "5cf016743caf73157c995336",
            telephone: "0850300073"
          },
          time: 0,
          change: 100
        }
      }
    });
    //expect status 1 because 0 is available and 1 is unavailable
    expect(costReduce.cost.status).toEqual("0");
    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_LOCKER_SELECTED");
  });
});

describe("Test Case 5", () => {
  let store;
  let costReduce;

  beforeEach(() => {
    // mock data
    store = mockStore({
      locker: [],
      cost: { cost: {}, money: 0 },
      modal: { successModal: false, unAvailableModal: false, change: 0 },
      loading: false
    });
  });

  test("renders", () => {
    // check component has exists
    const wrapper = shallow(
      <Provider store={store}>
        <LockerComponent />
      </Provider>
    );

    expect(wrapper.exists()).toBe(true);
  });

  test("Request Locker", async () => {
    // mock api
    await store.dispatch(mockLockers());
    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_LOCKER_SELECTED");
  });

  test("Set money and select locker", async () => {
    // mock insert cash 100
    await store.dispatch(setMoney(400));
    // mock locker
    await store.dispatch(setCost());
    const actions = store.getActions();
    const costs = cost(undefined, {
      type: "UPDATE_MONEY",
      payload: {
        money: 400
      }
    });
    // expect insert cash 100 success
    expect(costs.money).toEqual(400);
    // mock locker
    costReduce = cost(undefined, {
      type: "UPDATE_COST",
      payload: {
        cost: {
          _id: "5ceea163de8adf2ef81d502b",
          locker: 12,
          size: {
            _id: "5cee905b5162c31b700a185a",
            size: "L",
            perhour: 100,
            nextminute: 150
          },
          income: 700,
          status: "0",
          timeout: "2019-06-02T18:49:29.166",
          user: {
            _id: "5cf016743caf73157c995336",
            telephone: "0850300073"
          },
          time: 145,
          change: 0
        }
      }
    });
    //expect time is 60 minutes
    expect(costReduce.cost.time).toEqual(145);
    //expect charge is 50 bath perhour
    expect(costReduce.cost.size.perhour).toEqual(100);
    //expect next minute
    expect(costReduce.cost.size.nextminute).toEqual(150);
    //expect total perhour and nextminute
    const total = parseInt(costReduce.cost.size.perhour * 4);
    expect(total).toEqual(400);
    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MONEY");
    //expect action redux
    expect(actions[1].type).toEqual("UPDATE_COST");
  });

  test("Confirm Order And Rent Success", async () => {
    //mock open success modal
    await store.dispatch(setStatusSuccessModal(true));
    //call function calculateChange like cash and coin
    const change = calculateChangeCashAndCoin(costReduce.cost.change).filter(
      cash => cash !== 0
    );
    //expect not receiving change
    expect(change.length).toEqual(0);

    await store.dispatch(setChangeToSuccessModal(0));
    await store.dispatch(deleteMoney());

    const actions = store.getActions();

    //expect action redux
    expect(actions[0].type).toEqual("UPDATE_MODAL_SUCCESS");
    expect(actions[1].type).toEqual("UPDATE_CHANGE_SUCCESS");
    expect(actions[2].type).toEqual("DELETE_MONEY");
  });
});

// this function for calculate cash and coin.
const calculateChangeCashAndCoin = change => {
  let cash = [1000, 500, 100, 50, 20, 10, 5, 2, 1];
  let changeCash = [];

  if (change > 0) {
    cash.forEach((value, index) => {
      const getCash = parseInt(cash[index]);
      changeCash[index] = parseInt(change / getCash);
      change = change - changeCash[index] * getCash;
    });
  }

  return changeCash;
};
