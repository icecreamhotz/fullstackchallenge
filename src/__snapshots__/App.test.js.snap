// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Case 1 Request Locker 1`] = `ShallowWrapper {}`;
